### ISSUE TITLE
Building an index of episode id's with corresponding URLs

### ISSUE THREAD

**Issue by RevanthRameshkumar** [RevanthRameshkumar/CRD3#5]
I found this API recently and thought this was a good way to get rid of any noise in the summaries that happened due to text scraping:
https://criticalrole.fandom.com/api/v1#!/Articles/getList_get_2

Briefly looking through the API functionality, it looks like we need to build an index of episode ID's so that we can use:
https://criticalrole.fandom.com/api/v1/Articles/AsSimpleJson?id=**<episode-id\>** to get the summaries.

It looks like we can use the episode titles to and the lexical offset feature to get the ID's, ex:
https://criticalrole.fandom.com/api/v1/Articles/List?limit=25&offset=Arrival_at_Kraghammer
This means you need the episode names, but we do have a neat list of them already!

It is kind of a roundabout way to get the ID's, so do let me know if you find a better method!

The expected outcome of this project (open to discussion 😄) would be a .py file that we can run that uses the API to generate an output episode ID tsv with the proposed fields:

episode name \t episode url \t episode campaign number \t episode number in campaign \t episode wiki ID

Once we have a stable way of getting ID's, we can rebuild the summary data and use a diff to see how much error we had in the original data due to web scraping (even with my regex based cleaning). If the new summary data is cleaner, we can replace the old data and perhaps rerun the alignment algorithms!

**Comment by lpmi-13** [comment]
:+1:  I'll take a look!

**Comment by lpmi-13** [comment]
okay, so looks like imdb API is a more reliable way to get the episode titles (https://imdbpy.readthedocs.io/en/latest/usage/series.html), though that doesn't eliminate the issue with the critical role wiki URLs for each episode...

Am I right in assuming the main reason we want to access a URL like https://criticalrole.fandom.com/wiki/Vorugal_(episode) is so we can access the transcript link (ie, https://crtranscript.tumblr.com/ep071)?

It's also very strange that IMDB lists a bunch of seasons (15 for some reason...), but at least the two campaigns we're interested in seem to be confined to the "first two seasons", so using the python wrapper to the IMDB API seems stable.

**Comment by lpmi-13** [comment]
Ah, so it also seems like using `https://criticalrole.fandom.com/wiki/EPISODE_TITLE_HERE/Transcript` is stable/predictable, and that could be one way to get around the "(episode)" issue you mentioned above.

So, for example, `https://criticalrole.fandom.com/wiki/Thar_Amphala/Transcript` resolves to a page with transcripts, whereas `https://criticalrole.fandom.com/wiki/Thar_Amphala_(episode)/Transcript` resolves to "not found" type of page. I'll go ahead and build URLs like that and we'll see where we end up.

**Comment by lpmi-13** [comment]
and it also looks like the Campaign and episode numbers are predictable (first episode of the season 2 list = 2x01), so I'm wondering if we even need the episode wiki ID or episode URL?

so if we ended up with a tsv like:
`episode name \t episode transcript url \t episode campaign number \t episode number in campaign`
...would that be enough?

**Comment by lpmi-13** [comment]
got something worked up and pushed to my fork on the `adding_scripts` branch:

https://github.com/lpmi-13/CRD3/blob/adding_scripts/imdb_access.py
https://github.com/lpmi-13/CRD3/blob/adding_scripts/output.tsv

lemme know if this is closer to what we need.

**Comment by RevanthRameshkumar** [comment]
Hey that looks pretty good so far! Good catch on the transcript page being resolved properly (and so we can backtrack into the actual content page). I think one thing that we still need to figure out is how we get the internal episode ID to properly use this api call:
https://criticalrole.fandom.com/api/v1/Articles/AsSimpleJson?id=<episode-id>
This is an inefficient idea, but I wonder if we can use the lexicographic search using that other api call w/ the episode names and see which of the resulting url's match up w/ your imdb provided urls. From there you can just grab that internal episode id from the api call result and make the above call for the actual episode text.

**Comment by lpmi-13** [comment]
Cool, I go ahead and work that up. Do you have a preferred spider/crawler? I'm usually partial to scrapy,  but it's probably a bit heavy for just "click on link and go back one page", so even a combo of beautifulsoup and requests is probably a lighter touch. Lemme know what you think.

**Comment by RevanthRameshkumar** [comment]
Beautifulsoup is probably fine :) Do let me know if you have a more efficient way to get to the episode id's though. It is strange to me that the api exposes the episode only through an internal id, then makes the internal id hard to get.

**Comment by lpmi-13** [comment]
so unfortunately, it also doesn't look like the transcript pages have predictable back links.

ie, https://criticalrole.fandom.com/wiki/Thar_Amphala/Transcript

now has a link that leads back to

https://criticalrole.fandom.com/wiki/Thar_Amphala_(city)

...there is technically text that helps us get to where we want to go, but now this is seeming like a supremely hacked out solution that is not likely to operate in a deterministic fashion in the future (I assume because there's no central authority maintaining naming/linking conventions).

![text to be detected and scraped](https://user-images.githubusercontent.com/5070516/92588543-11246900-f291-11ea-9f0f-d0e578060690.png)

Anyway, I'm leaning towards just compiling a list of all the IDs and just updating that as/when new shows come out, particularly since every other thing we seem to need for the output `tsv` is (at least for now) deterministic.

What would you think about having a list like `episode-ids.txt` checked into version control alongside the rest of this?

**Comment by RevanthRameshkumar** [comment]
> ...there is technically text that helps us get to where we want to go, but now this is seeming like a supremely hacked out solution that is not likely to operate in a deterministic fashion in the future (I assume because there's no central authority maintaining naming/linking conventions).

totally agree

> What would you think about having a list like episode-ids.txt checked into version control alongside the rest of this?,

 yeah i think that works..or just put the id's as another column in the existing tsv

**Comment by lpmi-13** [comment]
alright, have pushed an updated version of `output.tsv` to the same `added_scripts` branch, with the episode number now included. I can also add some of the intermediate scripts if that would be helpful. Essentially, the steps were:

- get all the episode names from the imdb API
- standardize the episode names to turn spaces to underscores and get rid of "..."
- use these names to get the ID from the "offset" endpoint
- test those IDs to see if they match with a json result that has some descriptive text about an episode, assumed to be a marker of the correct ID for an episode
- collect all the IDs that don't lead to this text, and output those to a file
- test the episode names and IDs in that "bad id" file manually, and update (the two usual issues were adding the `(episode)` to the name like you mentioned, and for some reason all of the URLs with apostrophes work except for `https://criticalrole.fandom.com/wiki/K'varn_Revealed/Transcript`, so that had to be manually switched to use the HTML encoding `%27` instead of the apostrophe).
- ...and at the end of all that, the names and episodes can be verified and added to the tst output, as I've done.

**Comment by RevanthRameshkumar** [comment]
This is awesome! It looks like you did very little manual work which is great!
Do you want to make a PR? And please do include those scripts :)...all I ask is that they follow the linting if that is no bother
Once that is in, the next step would be either to fill in the missing transcripts (for the episodes past 159) or update the summaries for the current episodes using the much better json output from the API. After both of those done, we can try running the NW algorithm on the refreshed set of episodes.
I'm currently working on some dialogue analysis, unrelated to summarization which I can check in in a couple of weeks :D

**Comment by lpmi-13** [comment]
Yep, I'll raise a PR with that today.

**Comment by lpmi-13** [comment]
oh right, and looks like we've got black/autopep8/flake8 in the requirements/github workflows, but only flake8 is getting run in the pipeline. I assume just linting the same way with flake8 would be sufficient, but lemme know if you had something else in mind.

**Comment by lpmi-13** [comment]
PR opened https://github.com/RevanthRameshkumar/CRD3/pull/11

**Comment by lpmi-13** [comment]
Oh, and please feel free to push back on any of this stuff...I'm not super satisfied with my naming of things, but figured it would be faster to just get your feedback.  I'm very fine with changing any of this in any way that you think makes more sense.

**Comment by RevanthRameshkumar** [comment]
Thanks for the PR! Will review it over the weekend. Maybe let's move it to the src folder.
Also I think I setup the workflow incorrectly because it is not triggering for the PR...perhaps because it is from a forked repo? I'll try to fix that first.

**Comment by lpmi-13** [comment]
Ah yes...I believe github actions are off by default for forks. I can run
it on mine if you'd like to see the outcome prior to merging

**Comment by RevanthRameshkumar** [comment]
Is it possible to just clone the repo and make a PR that way? I think that would trigger the workflows

**Comment by lpmi-13** [comment]
ah, it looks like commit did trigger the workflow, which is available here:

https://github.com/lpmi-13/CRD3/actions/runs/261377383

**Comment by RevanthRameshkumar** [comment]
oh interesting, so i guess we can just check the commits in the forked repo. Weird that github wouldn't display them in the repo that you want to merge into.
Ok I'll do a review after work, probably the first thing to do is to move the code into the src folder (maybe under utils?) so the linter checks it and the data into the data folder. I'll make a test or 2 and merge w/ that branch in the meantime

**Comment by lpmi-13** [comment]
Yeah, now that I think about it, I've definitely seen actions run in other
repos against PRs I've opened from forks in the past.

I'll go ahead and make those changes and rebase the PR, and maybe tinker
around a bit to see if I can figure out what's going on.

**Comment by RevanthRameshkumar** [comment]
hey it seems to work now:
https://github.com/RevanthRameshkumar/CRD3/pull/14

I think I didn't add the pull_request tag in the workflow trigger, but after I added it there was no change to your PR so there was no event for trigger

**Comment by lpmi-13** [comment]
ah right...now I remember. I haven't really used github actions that much, so still a bit hazy on the details. I'll rebase and push again and it should fire the action.

**Comment by RevanthRameshkumar** [comment]
Hey, sorry I've been slow on this! Lot's of work stuff came up, overall looks good...I think some commenting would be good. I'll take a crack at that in the morning and PR into your branch.

**Comment by RevanthRameshkumar** [comment]
All done. Sorry I accidently pushed into your branch somehow w/o making a PR (didn't know I could do that on GitHub without permissions). https://github.com/lpmi-13/CRD3/tree/adding_scripts
I did some trivial rearranging too. If you think everything looks fine, remake the PR with these changes and we can get it in.

**Comment by lpmi-13** [comment]
Yeah, PRs are their own special thing in github, and aren't technically on a branch, but enough with the boring stuff, these updates look awesome! I can rebase them on my PR so the history looks nice, or you can just merge in as is. Up to you.

**Comment by RevanthRameshkumar** [comment]
I think I got it working now w/ the pull_request_review flag (check out the workflow to see what I mean).
Merged your stuff w/ rebase, thanks for this great work!
Will close this issue and make a couple of new ones later today if you want to tackle more. I think the 3 main ones will be
1. fill in the transcripts for the new episodes not in the data folder
2. update the summaries for all episodes using the wiki instead of my html parsing
3. rerun NW algorithm on the new transcript summary pairs (I'll finish up the post-processing and testing in the nw algorithm file so people can use it).

once we are done with that, we can start branching off into experiments like testing new scoring functions for NW algorithm (currently I use a scaled ROUGE-F1) or even trying to build a statistical alignment model.
Of course, we can do things not related to summarization as well :D, I've been meaning to explore dialogue acts and game flow...so many we can work on a tool that resolves dice roll requests and actions? I think that would be really neat training data for game flow.
